tests/integration_tests/modules/test_hotplug.py

```diff
@@ -394,3 +395,145 @@ def test_no_hotplug_triggered_by_docker(client: IntegrationInstance):
+        f" {result.return_code}\nstdout: {result.stdout}\n"
+        f"stderr {result.stderr}"
+    )
+
+
+BLOCK_CLOUD_CONFIG = """\
+[Unit]
+Description=Block cloud-config.service
+After=cloud-config.target
+Before=cloud-config.service
+
+DefaultDependencies=no
+Before=shutdown.target
+Conflicts=shutdown.target
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/sleep 360
+TimeoutSec=0
+
+# Output needs to appear in instance console output
+StandardOutput=journal+console
+
+[Install]
+WantedBy=cloud-config.service
+"""  # noqa: E501
+
+
+BLOCK_CLOUD_FINAL = """\
+[Unit]
+Description=Block cloud-final.service
+After=cloud-config.target
+Before=cloud-final.service
+
+DefaultDependencies=no
+Before=shutdown.target
+Conflicts=shutdown.target
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/sleep 360
+TimeoutSec=0
+
+# Output needs to appear in instance console output
+StandardOutput=journal+console
+
+[Install]
+WantedBy=cloud-final.service
+"""  # noqa: E501
+
+
+def _customize_environment(client: IntegrationInstance):
+    def enable_block_service(service_name: str, content: str):
+        service_filename = f"/lib/systemd/system/{service_name}.service"
+        client.write_to_file(service_filename, content)
+        client.execute(f"chmod 0644 {service_filename}", use_sudo=True)
+        client.execute(
+            f"systemctl enable {service_name}.service", use_sudo=True
+        )
+
+    enable_block_service("block-cloud-config", BLOCK_CLOUD_CONFIG)
+    enable_block_service("block-cloud-final", BLOCK_CLOUD_FINAL)
+
+    # Allows ssh access for 1000 user before boot has finished
+    contents = client.read_from_file("/etc/pam.d/sshd")
+    contents = (
+        "account [success=1 default=ignore] pam_succeed_if.so quiet uid eq"
+        " 1000\n\n" + contents
+    )
+    client.write_to_file("/etc/pam.d/sshd", contents)
+
+    client.instance.shutdown(wait=True)
+    client.instance.start(wait=False)
+
+
+@pytest.mark.skipif(
+    PLATFORM != "ec2",
+    reason="test is ec2 specific but should work on other platforms with the"
+    " ability to add_network_interface",
+)
+@pytest.mark.user_data(USER_DATA)
+def test_nics_before_config_trigger_hotplug(client: IntegrationInstance):
+    """
+    Test that NICs added/removed after the Network boot stage but before
+    the rest boot stages do trigger cloud-init-hotplugd.
+
+    Note: Do not test first boot, as cc_install_hotplug runs at
+    config-final.service time.
+    """
+    _customize_environment(client)
+
+    # wait until we are between cloud-config.target done and
+    # cloud-config.service
+    wait_for_cmd(client, "systemctl status cloud-config.target", 0)
+    wait_for_cmd(client, "systemctl status block-cloud-config.service", 3)
+
+    assert_systemctl_status_code(client, "cloud-config.service", 3)
+    assert_systemctl_status_code(client, "cloud-final.service", 3)
+
+    added_ip_0 = client.instance.add_network_interface()
+    _wait_till_hotplug_complete(client, expected_runs=1)
```

Comment:
Before we `_wait_till_...`, I think that we should assert that we are still between `cloud-config.target` done and `cloud-config.service`? If not, then this test didn't do what we thought it did, right?

Reply:
What I am currently assert is that `cloud-config.target` is finished in line [514: wait_for_cmd(client, "systemctl status cloud-config.target", 0)](https://github.com/canonical/cloud-init/pull/5058/files/6ee132506d6b966e80ee2e2c54639d3bbe6c55af#diff-fe6975e54f120fbe110935bb6e5cd8ba9b5dfe6613eae48787bcbc079c582299R514) and that `cloud-config.service` has not yet started in line  [517: assert_systemctl_status_code(client, "cloud-config.service", 3)](https://github.com/canonical/cloud-init/pull/5058/files/6ee132506d6b966e80ee2e2c54639d3bbe6c55af#diff-fe6975e54f120fbe110935bb6e5cd8ba9b5dfe6613eae48787bcbc079c582299R517). I think that is exactly what you mean, or am I missing something?

Reply:
Those assertions tell us before we add the interface that cloud-config.target is finished and that cloud-config.service has not yet started, but I think that a race is possible. If adding the network interface has some pre-flight check that takes a long time (or later adds something like this on accident), wouldn't it be possible that `add_network_interface()` doesn't actually add the interface until after cloud-config.service has started?

Reply:
With a 6 minute sleep this is all unlikely to happen, but I think that an assertion after adding the nic that we're still in the state we thought we were in would assert that the test is testing what we think it is.

Reply:
Gotcha, updated.